neqo-transport/src/streams.rs

```diff
@@ -0,0 +1,464 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Stream ID and stream index handling.
+
+use crate::fc::{
+    LocalStreamsFlowControls, ReceiverFlowControl, RemoteStreamLimits, SenderFlowControl,
+};
+use crate::frame::Frame;
+use crate::packet::PacketBuilder;
+use crate::recovery::RecoveryToken;
+use crate::recv_stream::{RecvStream, RecvStreams};
+use crate::send_stream::{SendStream, SendStreams, TransmissionPriority};
+use crate::stats::FrameStats;
+use crate::stream_id::{StreamId, StreamType};
+use crate::tparams::{self, TransportParametersHandler};
+use crate::ConnectionEvents;
+use crate::{Error, Res};
+use neqo_common::{qtrace, qwarn, Role};
+use std::cell::RefCell;
+use std::rc::Rc;
+
+pub struct Streams {
+    role: Role,
+    tps: Rc<RefCell<TransportParametersHandler>>,
+    events: ConnectionEvents,
+    flow_control_sender: Rc<RefCell<SenderFlowControl<()>>>,
+    flow_control_receiver: Rc<RefCell<ReceiverFlowControl<()>>>,
+    remote_streams_fc: RemoteStreamLimits,
+    local_streams_fc: LocalStreamsFlowControls,
+    pub(crate) send_streams: SendStreams,
+    pub(crate) recv_streams: RecvStreams,
+}
+
+impl Streams {
+    pub fn new(
+        tps: Rc<RefCell<TransportParametersHandler>>,
+        role: Role,
+        events: ConnectionEvents,
+    ) -> Self {
+        let limit_bidi = tps
+            .borrow()
+            .local
+            .get_integer(tparams::INITIAL_MAX_STREAMS_BIDI);
+        let limit_uni = tps
+            .borrow()
+            .local
+            .get_integer(tparams::INITIAL_MAX_STREAMS_UNI);
+        let max_data = tps.borrow().local.get_integer(tparams::INITIAL_MAX_DATA);
+        Self {
+            role,
+            tps,
+            events,
+            flow_control_sender: Rc::new(RefCell::new(SenderFlowControl::new((), 0))),
+            flow_control_receiver: Rc::new(RefCell::new(ReceiverFlowControl::new((), max_data))),
+            remote_streams_fc: RemoteStreamLimits::new(limit_bidi, limit_uni, role),
+            local_streams_fc: LocalStreamsFlowControls::new(role),
+            send_streams: SendStreams::default(),
+            recv_streams: RecvStreams::default(),
+        }
+    }
+
+    pub fn client_0rtt_rejected(&mut self) {
+        self.send_streams.clear();
+        self.recv_streams.clear();
+        self.remote_streams_fc = RemoteStreamLimits::new(
+            self.tps
+                .borrow()
+                .local
+                .get_integer(tparams::INITIAL_MAX_STREAMS_BIDI),
```

Comment:
Do we really need to update the remote stream limits here?  We won't have read anything from the server that would have caused us to change the limits.  Would a (debug) assertion that they haven't change suffice?

Reply:
I will do this in a follow up. This PR is trying not to change behavior, just to move things around.